runtime/src/chpl-privatization.c

```diff
@@ -41,30 +53,25 @@ static inline int64_t max(int64_t a, int64_t b) {
 // elements. Be __very__ careful if you have to update it.
 void chpl_newPrivatizedClass(void* v, int64_t pid) {
   chpl_sync_lock(&privatizationSync);
```

Comment:
I don't want this PR to change this, but do you think we could change chpl_newPrivatizedClass / chpl_clearPrivatizedClass to no longer take a lock, and instead use an atomic? But related to this PR: how do we know we're not updating an "old" version of the array, in the event that it is resized? Is that currently relying on the lock? (It seems clear to me that the lock prevents it - I'm asking if something else would, if the lock were per-element, say).

Reply:
Yes, unfortunately some synchronization is required in this case. The indirection of using recycling data in blocks would allow for both non-resizing `chpl_newPrivatizedClass` and `chpl_clearPrivatizedClass` to proceed without synchronization, but then `chpl_getPrivatizedClass` would suffer a lot.

One alternative is a reader-writer lock, as both non-resizing `chpl_newPrivatizedClass` and `chpl_clearPrivatizedClass` can count as readers and only resizing `chpl_newPrivatizedClass` would need to upgrade to a writer (either with our own upgrade-able Reader-Writer locks built on top of sync variables or `pthread_rwlock_t`). Furthermore, contention can be eliminated by using the 'distributed lock' concept I entertained before by having each thread have its own thread-local reader-writer lock that a single writer must acquire (either in some locking order or by acquiring some 'master' lock). This would significantly improve performance I believe.